framework/src/task_manager/task_manager_core.c

```diff
@@ -556,59 +551,146 @@ int taskmgr_get_handle_by_pid(int pid)
 		return;
 	}
-	msg_broad.sival_int = msg;
+
 	for (handle = 0; handle < CONFIG_TASK_MANAGER_MAX_TASKS; handle++) {
 		if (TM_LIST_ADDR(handle) != NULL) {
 			ret = taskmgr_get_task_state(handle);
 			if (ret == TM_APP_STATE_STOP || ret == TM_APP_STATE_UNREGISTERED) {
 				continue;
 			}
-			ret = taskmgr_check_msg_mask(msg, handle);
-			if (ret != OK) {
+			ret = taskmgr_get_tm_broad_msg(msg);
+			broadcast_info = taskmgr_search_broadcast_info(msg, handle);
+			if (broadcast_info == NULL) {
 				continue;
 			}
 			ret = ioctl(fd, TMIOC_BROADCAST, TM_PID(handle));
 			if (ret != OK) {
 				continue;
 			}
+			msg_broad.sival_ptr = broadcast_info;
 			(void)sigqueue(TM_PID(handle), SIGTM_BROADCAST, msg_broad);
 		}
 	}
 }
 
+static void taskmgr_broadcast_msg_init(void)
+{
+	int chk_idx;
+	int msg_num;
+
+	for (chk_idx = 0; chk_idx < -1 * TM_BROADCAST_PRE_DEFINE; chk_idx++) {
+		msg_num = chk_idx + 1;
+		tm_broadcast_msg[chk_idx].input_msg = -1 * (msg_num);
+		tm_broadcast_msg[chk_idx].tm_broad_msg = msg_num;
+	}
+}
+
 static int taskmgr_set_msg_cb(int type, void *data, int pid)
 {
 	int handle;
 
 	if (data == NULL) {
 		return TM_INVALID_PARAM;
 	}
-
 	handle = taskmgr_get_handle_by_pid(pid);
 	if (handle == TM_UNREGISTERED_APP) {
 		return handle;
 	}
-
 	if (type == TYPE_UNICAST) {
 		TM_UNICAST_CB(handle) = (_tm_unicast_t)data;
 	} else {
-		TM_MSG_MASK(handle) = ((tm_broadcast_t *)data)->msg_mask;
-		TM_BROADCAST_CB(handle) = ((tm_broadcast_t *)data)->cb;
+		if ((TM_BROADCAST_PRE_DEFINE != 0) && (tm_broadcast_msg[0].tm_broad_msg != 1)) {
+			taskmgr_broadcast_msg_init();
+		}
+		tm_broadcast_t *broadcast_info = (tm_broadcast_t *)TM_ALLOC(sizeof(tm_broadcast_t));
+		((tm_broadcast_t *)broadcast_info)->flink = NULL;
+		((tm_broadcast_t *)broadcast_info)->msg = taskmgr_get_tm_broad_msg(((tm_broadcast_t *)data)->msg);
+		((tm_broadcast_t *)broadcast_info)->cb = ((tm_broadcast_t *)data)->cb;
+		((tm_broadcast_t *)broadcast_info)->cb_data = ((tm_broadcast_t *)data)->cb_data;
+		sq_addlast((FAR sq_entry_t *)broadcast_info, &TM_BROADCAST_INFO(handle));
 	}
 	return OK;
 }
 
+static int taskmgr_set_new_broadcast_msg(int msg)
+{
+	int chk_idx;
+	int msg_num;
+
+	if ((TM_BROADCAST_PRE_DEFINE != 0) && (tm_broadcast_msg[0].tm_broad_msg != 1)) {
+		taskmgr_broadcast_msg_init();
+	}
+	for (chk_idx = 0; chk_idx < CONFIG_TASK_MANAGER_MAX_TASKS; chk_idx++) {
+		msg_num = chk_idx + 1;
+		if (tm_broadcast_msg[chk_idx].input_msg == msg) {
+			return OK;
+		} else {
+			if (tm_broadcast_msg[chk_idx].tm_broad_msg != msg_num) {
+				tm_broadcast_msg[chk_idx].input_msg = msg;
+				tm_broadcast_msg[chk_idx].tm_broad_msg = msg_num;
+				return OK;
+			}
+		}
+	}
+	return TM_OUT_OF_MEMORY;
+}
+
+static int taskmgr_unset_broadcast_cb(int msg, int pid)
+{
+	int handle;
+	tm_broadcast_t *broadcast_info;
+
+	handle = taskmgr_get_handle_by_pid(pid);
+	broadcast_info = taskmgr_search_broadcast_info(msg, handle);
+	if (broadcast_info == NULL) {
+		return TM_INVALID_PARAM;
```

Comment:
this case is that we cannot find the broadcast info. is it a invalid param problem?